Features/constraints

thomasloux opened the pull request:
## Summary

First draft for constraints. Inspiration from ASE.
This is a rather big feature, I prefer to post a first mvp so that we can discuss some implementation and make sure constraints are compatible with Integrators and Optimizers.

### Codes implementing constraints
- JAX-MD: no
- ASE: yes
- LAMMPS: yes, but difficult to read (C++) and copy.
- OpenMM: yes, but only bonds lengths or angles as far as I understand
For the rest I did not check this much.

### Implementation:
- Constraints: FixAtoms, FixCom
Change in the code:
- Add an optional constraints variable in SimState
- Turn every subclass dataclass(kw_only=True) to prevent errors (positional argument before non pos argument)
- Adapt MD only, via setter (set_momenta, set_positions)

### What's working perfectly
- Contraints tests implemented
- All tests pass
- FixAtoms really fixes the atoms concerned

### Future work
- Adapt optimizer
- Make sure that FixCom is working well. The Com drift is small, but I would expect an even smaller value. ASE seems to have a smaller value from first tests. I did not yet reproduce their tests as it is based on an optimization task.
- Implement other constraints
- make sure everything is compatible with Integrators and Optimizers. Essentially we should apply RATTLE algorithm.
- Adapt the calculation of temperature to take into account the reduction of degrees of freedom.
- For FixCom,  one wants the center of mass to be fixed in unwrap coordinate. So of course, we need to check after unwrapping. And as a result, I now affect set_momenta, then apply the wrapping on the systems. Actually ASE does not wrap during the simulations, but the wrapping is applied from each calculator forward pass. I think that it is the case for most of TorchSim model implementations. So we may want to stop wrapping during the simulations. Of course if always possible to wrap afterwards.

### Implementation discussions
- I wanted first to have an hidden variable _positions and have a implicit setter and getter. But this is very painful to adapt with _atom_attributes and the copy of state sometimes performed directly accessing the variables using vars(state).  At least it is the case if I set the __init__ function to accept an `positions` argument and not a `_positions` argument. Actually I think that the setter is clearer to indicate that the constraints are imposed.
- I decided to add constraints as a global attribute, but prevent list of constraint and looping over the constraints.
- The disavantage: It's rather difficult to write the constraints and define the constraints over the batch system. For FixAtoms you need to provide the index of the atoms in the batch system. Probably we want to define actually for each system and then batch the constraint. For FixCom, we need to compute the COM in an efficient, which is not super easy to read. It will be even worse if one wants to fix the com of a subgroup in a batch systems. 

## Checklist

Before a pull request can be merged, the following items must be checked:

* [ ] Doc strings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google).
* [X] Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [X] Tests have been added for any new functionality or bug fixes.

We highly recommended installing the `prek` hooks running in CI locally to speedup the development process. Simply run `pip install prek && prek install` to install the hooks which will check your code before each commit.

curtischong reviewed (commented):
This looks good!

I like how the abstract methods like adjust_positions are flexible enough for constraining specific indices/atoms and for constraining specific axes of movement (which'll probably be added later)

orionarcher commented on torch_sim/constraints.py:
```diff
@@ -0,0 +1,406 @@
+        Returns:
+            Dictionary representation of the constraint
+        """
+        return {"name": self.__class__.__name__, "kwargs": self.__dict__.copy()}
```
I'd get rid of this too. `asdict` is a built-in for `dataclasses` and users can add a name themselves if they want.

thomasloux commented:
Hey, I've changed quite a lot the API of the constraints. Actually I've almost reproduced the equivalent of atom_attribute or system_attribute at the level of constraints. The code is still a draft, they are probably better way reorganize the operation, especially for manipulation of states (pop, split, concatenante) which are probably difficult to read and too long.

- Now support states manipulations
- I removed wrap
- Add a fix reference in FixCom (set at the first call)
- Test for constraints for optimizers
- Modify calc_kt to support new degrees of freedom

Regarding SystemConstraint, I need to implement a default slice(None) so that the constraint can function in the case the final user directly sets constraint (state.constraint = [FixCom()]). We could accept not to support this.

Also I realize the general_attribute is only supported for now for pbc, which should change soon. We should have constraint in its own category as it's not even propagate in similar ways as the rest of the attributes. This would also allow to have a private _constraint variable, and support the previous statement not to set directly constraint.

More than open for suggestion how to best add constraints.

By the way, in the meantime, we may want to have a dedicate branch so that people can try (a lot of demand for optimization #114). It's probably better to do before having extensive tests and proofs that our optimizers and integrators are compatible with the constraints. 

Regarding calc_kt, I suggest to make it a MDState method (like in ASE). So that the user does not forget to add degrees of freedom.

I did not add for now a check of incompatible constraints for now, not sure it is relevant. For instance you may want in the future to fix the length of all atoms in a chain (think of a polymer). Then the index of an atom will appear multiple times.

janosh commented:
@thomasloux great work on this! i'd like to start using symmetry constrained relaxation and would like to help push this over the finish line. if you don't mind i'll push a few minor bug fixes and some more tests?

thomasloux commented:
@janosh I've made the changes I wanted to do. Now everything  is working as I wanted. Let me know if you have questions or remark on my implementation

curtischong commented on torch_sim/constraints.py:
```diff
@@ -0,0 +1,470 @@
+    def get_removed_dof(self, state: SimState) -> torch.Tensor:
+        """Get number of removed degrees of freedom.
+
+        Each fixed atom removes 3 degrees of freedom (x, y, z motion).
```
It'd be nice if people that implemented this function wouldn't have to think about systems at all. they just think about how many dofs it removes per system

orionarcher replied:
Yeah I don't think there is really any way around needing to be aware of systems, nor is it a big burden on constraint developers imo

curtischong commented on torch_sim/constraints.py:
```diff
@@ -0,0 +1,470 @@
+        return f"FixAtoms(indices={indices_str})"
+
+
+class FixCom(SystemConstraint):
```
I think this is only valid for non PBC systems right? Maybe we should throw a warning if they're doing it on a state with pbc?

curtischong commented on torch_sim/models/einstein.py:
```diff
@@ -0,0 +1,285 @@
+
+Contrary to other models, the model energies depend on an absolute reference position,
+so the model can only be used on systems that the model was initialized with.
+As a analytical model, it can provide its Helmholtz free energy and can also generate
```
nit: "an analytical"

curtischong commented on torch_sim/models/einstein.py:
```diff
@@ -0,0 +1,285 @@
+        beta = 1.0 / (kB * temperature)  # Inverse temperature in 1/eV
+
+        # Sample positions from a normal distribution around equilibrium positions
+        stddev = torch.sqrt(1.0 / (self.masses * (self.frequencies**2) * beta)).unsqueeze(
```
maybe we can replace `self.masses * (self.frequencies**2)` with self.spring_constants

thomasloux replied:
This file is not supposed to be in this PR sorry. But this remark is relevant yes

curtischong commented on torch_sim/optimizers/fire.py:
```diff
@@ -211,13 +212,15 @@ def _vv_fire_step[T: "FireState | CellFireState"](  # noqa: PLR0915
     # Get new forces and energy
     model_output = model(state)
-    state.forces = model_output["forces"]
+    # state.forces = model_output["forces"]
```
same here

curtischong commented on torch_sim/optimizers/fire.py:
```diff
@@ -454,16 +457,19 @@ def _ase_fire_step[T: "FireState | CellFireState"](  # noqa: C901, PLR0915
     # Get new forces, energy, and stress
     model_output = model(state)
-    state.forces = model_output["forces"]
+    # state.forces = model_output["forces"]
```
this one too

curtischong commented on torch_sim/optimizers/state.py:
```diff
@@ -23,6 +23,39 @@ class OptimState(SimState):
+        _constraints: list | None = None,
+    ) -> None:
+        """Initialize optimization state."""
+        self.positions = positions
```
shouldn'y we do self.set_positions as well?

curtischong commented on torch_sim/constraints.py:
```diff
@@ -0,0 +1,470 @@
+        """
+        dtype = state.positions.dtype
+        n_systems = (
+            state.n_systems if self.system_idx == slice(None) else len(self.system_idx)
```
I do think it's fine since ik we were thinking of changing system_idx to a sparse tensor due to performance

curtischong commented on torch_sim/state.py:
```diff
@@ -235,7 +294,16 @@ def get_number_of_degrees_of_freedom(self) -> torch.Tensor:
+                removed_dof = constraint.get_removed_dof(self)
+                dof_per_system -= removed_dof
+
+        # Ensure non-negative DOF
```
Maybe we should warn if the degrees of freedom is below 0? Actually, maybe we can put it inside validate_constraints? I can see a case where the dof is 0 but if it's below 0 I think it's more often an error?

thomasloux replied:
actually even dof 0 is a problem. There is no motion then

curtischong commented on torch_sim/state.py:
```diff
@@ -649,8 +718,16 @@ def _filter_attrs_by_mask(
     Returns:
         dict: Filtered attributes with appropriate handling for each scope
     """
+    # atoms_mask = torch.isin(state.system_idx, torch.nonzero(system_mask).squeeze())
```
nit: unused comment

curtischong reviewed (commented):
There are definitely things I'll have to rereview after reading Orion's PR. But thank you for this PR! It's currently quite strong and brings us a good step towards implementing constraints

thomasloux commented on torch_sim/optimizers/state.py:
```diff
@@ -23,6 +23,39 @@ class OptimState(SimState):
+        self.energy = energy
+        self.set_forces(forces)
+        self.stress = stress
+        SimState.__post_init__(self)
```
indeed, changed to __post__init__ to enforce force constraints

curtischong reviewed (approved):
I think this is great! There are a few things I think could be improved but in general (e.g. allowing 0 dof in state.py), but it seems like we have constraints now.

curtischong commented on torch_sim/state.py:
```diff
@@ -238,6 +247,46 @@ def row_vector_cell(self, value: torch.Tensor) -> None:
         """
         self.cell = value.mT
 
+    def set_positions(self, new_positions: torch.Tensor) -> None:
```
I think this is a small but important thing to do because we don't want people to accidentally use the wrong functions to set positions/forces.

orionarcher replied:
Yeah I agree having a one liner is nice. What about: 
```python
from constraints import set_constrained_positions

set_constrained_positions(state, new_positions)
```
I prefer not having two ways to do the same thing and I like having all the constraints logic in constraints. Could call it something else, I'd be ok with simply `set_positions` too.

thomasloux replied:
I've changed my mind. One liner like ASE is a good option when you have a private variable _positions and force final user to use `set_positions`. In our case, it's not possible. So I think it's better as you said to keep in 2 lines
```
state.positions = positions
state.constrain_positions()
```
The code is more verbose but more self explainable as well.
I think I prefer this form as one understand that constraints come from state and modify internal positions, rather than  `constrain_positions` which I think is a bit weird either with the argument/signature `constrain_positions(new_positions, state.constraints, state)` or `constrain_positions(new_positions, state)`. In one case, it's a bit weird to provide state (but it's needed for some constraints (eg FixAtoms). The second case, the signature is the same as state.constraint_positions().

thomasloux commented on torch_sim/constraints.py:
```diff
@@ -0,0 +1,581 @@
+
+        # Ensure we have the right shape and type
+        system_idx = torch.atleast_1d(system_idx)
+        if system_idx.ndim > 1:
```
agreed on the if statement, but not on error message, although it's a really small detail
